specs/language/resources.tex

```diff
@@ -663,19 +664,241 @@
 
 \Sec{Constant Buffers}{Resources.cnbuf}
 
+Constant buffer represent a resource that contains read-only constant data in a specific memory layout.
+
+\Sub{Constant Buffer Declaration Block}{Resources.cnbuf.cb}
+
+\p Constant buffer can be declared using the \texttt{cbuffer} specifier.
+
+\begin{grammar}
+  \define{cbuffer-declaration-group}\br
+  \terminal{cbuffer} name \opt{resource-binding} \terminal{\{}
+  \opt{cbuffer-declaration-seq} \terminal {\}}
+
+  \define{cbuffer-declaration-seq}\br
+  \textit{cbuffer-declaration}\br
+  \textit{cbuffer-declaration-seq cbuffer-declaration}
+
+  \define{cbuffer-declaration}\br
+  \textit{variable-declaration}\br
+  \textit{namespace-declaration}\br
+  \textit{function-declaration}\br
+  \textit{record-declaration}\br
+  \textit{empty-declaration}
+\end{grammar}
+
+\p The name of the \texttt{cbuffer} declaration group cannot be referenced from within the translation unit and is not required to be unique.
+
+\p Variable declarations in the \texttt{cbuffer} declaration group are called \textit{shader constants}.
+
+\p Shader constants can be referenced from anywhere in the translation unit after they are declared by directly using the declaration name. This implies that all shader constants declared in a translation unit must have unique names, even though they might be declared in different \texttt{cbuffer} declaration groups.
+
+\p Variable declarations in the \texttt{cbuffer} declaration group cannot have \texttt{groupshared} or \texttt{static} variable modifiers.
+
+\p Any declaration in the \texttt{cbuffer} declaration group that is not a
+\textit{namespace-declaration} or a \textit{variable-declaration} is going to be
+processed as if it was declared outside of the group at the
+global scope.
```

Comment:
I'm trying to think about the relationship we want between namespaces and cbuffers. We got some GAB feedback about namespaces that referenced https://github.com/microsoft/DirectXShaderCompiler/issues/4514, which is wanting to put cbuffers in a namespace, and we similarly have this issue you filed which seeks to support namespaces inside a cbuffer https://github.com/microsoft/DirectXShaderCompiler/issues/7322.

My instinct is that we should make cbuffers only contain variable declarations and disallow namespaces in cbuffers, and instead allow cbuffers in namespaces.

The benefit of that approach is that we don't ever have to restrict what we can put in a namespace because it is enclosed in a cbuffer.

Thoughts?

Reply:
Oh! I think I missed that those issues are worded differently but refer to the same issue.

DXC does allow namespaces in cbuffers (and they work), but not cbuffers in namespaces. I feel like this relationship is backwards.

Reply:
DXC actually allows cbuffers in namespaces as long as it is not the only thing in the namespace, which is a bug we should fix: https://godbolt.org/z/GaWsPxxa7

I agree that we should make cbuffers only contain variable declarations. I'll update the spec.